src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject.cs

```diff
@@ -133,6 +135,29 @@ public virtual void Disconnect()
             VisualStudioProject.AddSourceFile(filename, sourceCodeKind, folders);
         }
 
+        protected void AddFile(
+            string filename,
+            string linkMetadata,
+            SourceCodeKind sourceCodeKind)
+        {
+            // We have tests that assert that XOML files should not get added; this was similar
+            // behavior to how ASP.NET projects would add .aspx files even though we ultimately ignored
+            // them. XOML support is planned to go away for Dev16, but for now leave the logic there.
+            if (filename.EndsWith(".xoml"))
+            {
+                return;
+            }
+
+            var folders = ImmutableArray<string>.Empty;
+            if (!string.IsNullOrEmpty(linkMetadata))
+            {
+                var linkFolderPath = Path.GetDirectoryName(linkMetadata);
```

Comment:
Has something validated that this is valid filename?  If not, I believe this will throw